src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs

```diff
@@ -101,36 +126,242 @@ CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> g
                             i == 0 ? openQuoteRange.End : interpolations[i - 1].CloseBraceRange.End,
                             interpolation.OpenBraceRange.Start)];
                         if (text.Length > 0)
-                        {
                             builder.Add(SyntaxFactory.InterpolatedStringText(MakeInterpolatedStringTextToken(text, kind)));
-                        }
 
                         builder.Add(ParseInterpolation(this.Options, originalText, interpolation, kind));
                     }
 
                     // Add a token for text following the last interpolation
                     var lastText = originalText[new Range(interpolations[^1].CloseBraceRange.End, closeQuoteRange.Start)];
                     if (lastText.Length > 0)
-                    {
                         builder.Add(SyntaxFactory.InterpolatedStringText(MakeInterpolatedStringTextToken(lastText, kind)));
+                }
+
+                CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> result = builder;
+                _pool.Free(builder);
+                return result;
+            }
+
+            CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> getMultiLineRawContent()
+            {
+                // If we have any errors in the multi-line literal, then don't bother to try to do fancy dedentation.
+                // There's no need as it's quite possible we don't even know what the dedent would be.
+                if (error != null)
+                    return getNormalContent();
+
+                // The indentation-whitespace computed from the very last line of the raw string literal
+                var indentationWhitespace = PooledStringBuilder.GetInstance();
+
+                // The leading whitespace of whatever line we are currently on.
+                var currentLineWhitespace = PooledStringBuilder.GetInstance();
+
+                // The content we want to create text token out of.  Effectively, what is in the text sections
+                // minus leading whitespace.
+                var content = PooledStringBuilder.GetInstance();
+                try
+                {
+                    var closeQuoteText = originalText[closeQuoteRange];
+
+                    // A multi-line raw interpolation without errors always ends with a new-line, some number of spaces, and the quotes.
+                    Debug.Assert(SyntaxFacts.IsNewLine(closeQuoteText[0]));
+
+                    var currentIndex = GetNewLineLength(closeQuoteText, index: 0);
+
+                    while (currentIndex < closeQuoteText.Length &&
+                        SyntaxFacts.IsWhitespace(closeQuoteText[currentIndex]))
+                    {
+                        indentationWhitespace.Builder.Append(closeQuoteText[currentIndex]);
+                        currentIndex++;
+                    }
```

Comment:
By saving the starting index, a single Append call can be made using the starting and (ending) current index after this loop.

Reply:
Yes. That would be much better

Reply:
\+ same at `currentLineWhitespace.Append` and `content.Append` a few more times below.

Is a builder still desired then, or could `indentationWhitespace` be passed around as a string or obtained from `closeQuoteText[startIndex..currentIndex]` or a `char[]`?